### ISSUE TITLE
rlp-gateway breaks after streaming for 15 minutes

### ISSUE THREAD

**Issue by gcapizzi** [cloudfoundry/loggregator#60]
After streaming logs from the `rlp-gateway` using the `RLPGatewayClient` for 15 minutes, we starting getting the following errors:

```
2018/11/26 15:17:28 unexpected status code 404: {"message":"resource not found","error":"not_found"}
```

We found this corresponding error in `reverse_log_proxy.stderr.log`:

```
2018/11/26 15:17:28.456878 Subscribe error: context canceled
```

We believe this is due to the connection between the gateway and the proxy timing out, but we couldn't find any 15 minutes timeout configured anywhere in the codebase. Possibly a gRPC default?

**Comment by bradylove** [comment]
Hey @gcapizzi,

We were able to reproduce the issue of the stream disconnecting. In our case we are testing with a GCP environment that was setup using BOSH Boot Loader and CF Deployment. When BOSH Boot Loader creates the load balancers for the environment it sets a timeout of 15 minutes. We need to do some further investigation on how to ensure that the GCP load balancer does not kill this connection.

On the issue of the 404's after the failure, it appears you are not updating the the auth token when requests are made with your auth client. By default in CF Deployment the tokens expire after 10 minutes. This can be resolved by updating your auth token any time [`Do`](https://github.com/cloudfoundry/cpu-entitlement-plugin/commit/7e87c24b544aecaf1e618ab524b4ce0d09a205a0#diff-03a4ce425e363e4656f842131ba995a0R78) is called.

**Comment by bradylove** [comment]
We were unable to find any way to keep the connection alive beyond the timeout configured in the load balancer so for a more graceful experience for the time being we have added a `closing` event to the stream sent by the RLP. This event is sent after the stream has been open for 14 minutes. We have then added logic to the client gracefully reconnect on a `closing` event.

See: https://github.com/cloudfoundry/loggregator/commit/237ebb991930d978375b7d9fb09c3e88531ba66a and https://github.com/cloudfoundry/go-loggregator/commit/7eacadc7c6c75fbf98bf9ed618e7d88fbb67ea44
